materialize-sql-json/driver.go

```diff
@@ -22,91 +27,95 @@ type Driver struct {
 	} else if resource, err = schemagen.GenerateSchema("SQL Table", d.ResourceSpecType).MarshalJSON(); err != nil {
 		return nil, fmt.Errorf("generating resource schema: %w", err)
 	}
 
-	return &pm.SpecResponse{
-		EndpointSpecSchemaJson: json.RawMessage(endpoint),
-		ResourceSpecSchemaJson: json.RawMessage(resource),
-		DocumentationUrl:       d.DocumentationURL,
+	return &protocol.SpecResponse{
+		DocumentationUrl:     d.DocumentationURL,
+		ConfigSchema:         json.RawMessage(endpoint),
+		ResourceConfigSchema: json.RawMessage(resource),
 	}, nil
 }
 
-// Validate implements the DriverServer interface.
-func (d *Driver) Validate(ctx context.Context, req *pm.ValidateRequest) (*pm.ValidateResponse, error) {
+func (d *Driver) Validate(ctx context.Context, req protocol.ValidateRequest) (*protocol.ValidateResponse, error) {
 	var (
-		err        error
-		endpoint   *Endpoint
-		loadedSpec *pf.MaterializationSpec
-		resp       = new(pm.ValidateResponse)
+		err            error
+		endpoint       *Endpoint
+		loadedBindings []protocol.ApplyBinding
+		resp           = new(protocol.ValidateResponse)
 	)
 
-	if err = req.Validate(); err != nil {
-		return nil, fmt.Errorf("validating request: %w", err)
-	} else if endpoint, err = d.NewEndpoint(ctx, req.EndpointSpecJson); err != nil {
+	if endpoint, err = d.NewEndpoint(ctx, req.Config); err != nil {
 		return nil, fmt.Errorf("building endpoint: %w", err)
-	} else if loadedSpec, _, err = loadSpec(ctx, endpoint, req.Materialization); err != nil {
+	} else if loadedBindings, _, err = loadBindings(ctx, endpoint, req.Name); err != nil {
 		return nil, fmt.Errorf("loading current applied materialization spec: %w", err)
 	}
 
 	// Produce constraints for each request binding, in turn.
 	for _, bindingSpec := range req.Bindings {
 		var constraints, _, resource, err = resolveResourceToExistingBinding(
 			endpoint,
-			bindingSpec.ResourceSpecJson,
-			&bindingSpec.Collection,
-			loadedSpec,
+			bindingSpec.ResourceConfig,
+			bindingSpec.Collection,
+			loadedBindings,
 		)
 		if err != nil {
 			return nil, err
 		}
 
 		resp.Bindings = append(resp.Bindings,
-			&pm.ValidateResponse_Binding{
+			protocol.ValidatedBinding{
+				ResourcePath: resource.Path(),
 				Constraints:  constraints,
 				DeltaUpdates: resource.DeltaUpdates(),
-				ResourcePath: resource.Path(),
-			})
+			},
+		)
 	}
 	return resp, nil
 }
 
-// ApplyUpsert implements the DriverServer interface.
-func (d *Driver) ApplyUpsert(ctx context.Context, req *pm.ApplyRequest) (*pm.ApplyResponse, error) {
+func (d *Driver) Apply(ctx context.Context, req protocol.ApplyRequest) (*protocol.ApplyResponse, error) {
+	// TODO(whb): For now we are only handling the case of deleting the entire materialization,
+	// which will be received as an ApplyRequest with no bindings. Future work could enable handling
+	// of deletion for individual bindings that are removed from a spec while the entire spec is not
+	// being deleted.
```

Comment:
I do remember the discussion about deleting individual bindings being problematic because other shards of the materialization might be running. I would think that all shards of the materialization would be stopped when doing an `Apply` and then re-started, otherwise the same problem would happen with adding bindings. Maybe that's the magic of handling deletes in `Apply` as well. But I'm not 100% sure either, so maybe @jgraettinger can comment here?

We did talk yesterday about deleting an entire materialization as an apply with no bindings. In that case it's connector-dependent on what they will do. Essentially I am advocating here that we continue to delete the tables from the endpoint as we would have when deleting materialization previously. That's helpful when running our own tests of the materialization, and also for users who are iterating on their specs.